File: backend/src/main/java/com/nusmotion/backend/dto/BusStop.java
package com.nusmotion.backend.dto;

import com.fasterxml.jackson.annotation.JsonProperty;

/**
 * Maps one bus stop from the upstream /BusStops response.
 *
 * LEARNING NOTE:
 * - Java records are ideal for DTOs: immutable, concise, equals/hashCode for free.
 * - @JsonProperty bridges the upstream PascalCase JSON to our camelCase fields.
 * - Jackson (the JSON library bundled with Spring Boot) uses these annotations
 *   during deserialization (JSON → Java) and serialization (Java → JSON).
 */
public record BusStop(
        @JsonProperty("caption") String caption,
        @JsonProperty("name") String name,
        @JsonProperty("LongName") String longName,
        @JsonProperty("ShortName") String shortName,
        @JsonProperty("latitude") double latitude,
        @JsonProperty("longitude") double longitude
) {}

File: frontend/lib/data/models/bus_stop.dart
class BusStop {
  final String caption;
  final String name;
  final String longName;
  final String shortName;
  final double latitude;
  final double longitude;

  const BusStop({
    required this.caption,
    required this.name,
    required this.longName,
    required this.shortName,
    required this.latitude,
    required this.longitude,
  });

  factory BusStop.fromJson(Map<String, dynamic> json) => BusStop(
    caption: json['caption'] as String? ?? '',
    name: json['name'] as String? ?? '',
    longName: json['LongName'] as String? ?? '',
    shortName: json['ShortName'] as String? ?? '',
    latitude: (json['latitude'] as num?)?.toDouble() ?? 0,
    longitude: (json['longitude'] as num?)?.toDouble() ?? 0,
  );
}
